feat(contact-us): create contact us component

nathanbowang opened the pull request.

nathanbowang committed:
feat(contact-us): create contact us component

nathanbowang committed:
feat(contact-us): add button hover style

Add button hover style.
Add input focus style.
Add textarea focus style.

nathanbowang committed:
feat(contact-us): add characters count

nathanbowang committed:
feat(contact-us): add asterisk to required field

nathanbowang committed:
style: update message form styles

nathanbowang committed:
style: update message form styles

nathanbowang committed:
style: update styles for webkit

reubenberghan commented on components/Main/ContactUs/ContactUs.module.css:
```diff
@@ -0,0 +1,17 @@
+.section {
```
This file seems to contain very similar styling / intent to the `NewsletterSignUp` component styles. Do they need to be separate?

nathanbowang replied:
I generalized a FormSection component. After these two branches merged, I will refactor NewsletterSignUp.

reubenberghan commented on components/Main/ContactUs/MessageForm/MessageForm.module.css:
```diff
@@ -0,0 +1,53 @@
+.form {
```
Again wondering what `form` and form control styles we can share across this and the `NewsletterSignUp` component?

reubenberghan commented on components/Main/ContactUs/MessageForm/index.tsx:
```diff
@@ -0,0 +1,68 @@
+  const NAME = 'name'
+  const EMAIL = 'email'
+  const BODY = 'body'
+  const MAX_BODY_LENGTH = 500
```
Do these `const`'s need to be defined in the body of the function component or could they be lifted up to the top of the file?

nathanbowang replied:
Either way works. I am wondering about the benefit of lifting it up to the top of the file. Do you mind explaining?

reubenberghan replied:
Yeah either way will work but in this case as they are not derived / dependent on the props it's not really necessary to have these defined here. This instance is more about reducing noise in the component.

The other thing to be aware of with React components is that each time this component is re-rendered this function is called so all of these values get redefined. Not much of a problem with a few strings like this but can run into issues with objects and functions. So generally it's a good practice to define any values / functions that are not derived from or have closure over props outside of the component.

nathanbowang replied:
That makes sense to me. I didn't consider that. Thanks.

reubenberghan commented on components/Main/ContactUs/MessageForm/index.tsx:
```diff
@@ -0,0 +1,68 @@
+  })
+
+  const handleChange = (event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
+    event.preventDefault()
```
Not sure we need this?

nathanbowang replied:
It is not useful here. `changeEvent.cancelable` is false.

reubenberghan commented on components/Main/ContactUs/MessageForm/index.tsx:
```diff
@@ -0,0 +1,68 @@
+      setMesage({
+        ...message,
+        [event.target.name]: event.target.value,
+        bodyLength: event.target.value.length,
```
Forms are always a fun one. Is the `bodyLength` relevant to the rest of the state or can it be managed separately?

reubenberghan commented on components/Main/ContactUs/MessageForm/index.tsx:
```diff
@@ -0,0 +1,68 @@
+      <label className={style.required} htmlFor={NAME}>
+        Name
+      </label>
+      <input id={NAME} name={NAME} onChange={handleChange} required type="text" value={message[NAME]} />
```
As `id`'s should be unique it might be better to scope these by using a template string. Perhaps something like `` `contact-message-form-${NAME}` ``?

reubenberghan commented on components/Main/ContactUs/index.tsx:
```diff
@@ -0,0 +1,21 @@
+import React, { useState } from 'react'
```
As mentioned above this does seem similar to the `NewsletterSignUp` component...

reubenberghan commented on components/Main/ContactUs/index.tsx:
```diff
@@ -0,0 +1,21 @@
+          <p>Your message has been sent successfully!</p>
+        ) : (
+          <MessageForm setIsSuccessful={setIsSuccessful}></MessageForm>
+        )}
```
Understand what you're doing here but worth considering whether it might be a better user experience if this wasn't a ternary. Yeah we should be showing a success message but should we be removing the form on success? Would it be reasonable to consider that someone might want to resubmit another message (or email for newsletter sign-up) once they have successfully submitted a response?

nathanbowang replied:
I steal the idea from [paperhat](paperhat.ltd).

reubenberghan replied:
Actually after reading the ACs I see that this is part of those so all good :grin:

reubenberghan reviewed (commented):
Nice work. A few things to discuss though :grin:

nathanbowang committed:
refactor: create a generic section component for forms

natebowang commented:
I haven't done all updates. I will push more commits and you could review after.

nathanbowang committed:
test(storybook): add stories to form section component

nathanbowang committed:
Merge branch 'demo' into feat/contact-us

nathanbowang committed:
test(storybook): adjust color contrast of the mock block component

Adjust color contrast of the MockBlock component to comply
accessibility. So the stories don't need the a11y selector.

nathanbowang committed:
refactor: create a generic input component for forms

nathanbowang committed:
refactor: create a generic button component for forms

nathanbowang committed:
refactor: update message form by using generic input and button

nathanbowang committed:
refactor: create a generic textarea component for forms

nathanbowang committed:
refactor: polish button, input, and textarea component

nathanbowang committed:
refactor: polish contact us component

nathanbowang committed:
feat(contact-us): add send another message button

nathanbowang committed:
refactor: create a generic form component for forms

nathanbowang committed:
style: refactor global.css

Refactor global.css.
Delete role attribute selector on list elements.

nathanbowang committed:
style: move max-width to section component

natebowang commented:
That's all updates I came up with so far. Please let me know if you have any further comments. Thanks.
